### PR TITLE
feat(core): users sensitive operation api

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

* feat(core): delete `/users/:userId`
* feat(core): patch `/users/:userId/password`
* feat(core): patch `/users/:userId/username`
* feat(core): patch `/users/:userId/primary-email`
* feat(core): patch `/users/:userId/primary-phone`

## Linear Issue Reference

* LOG-1261
* LOG-1260
* LOG-1259
* LOG-1251
* LOG-1250

## Testing

feat(core): delete `/users/:userId`
2-before-delete
2-delete
2-after-delete

feat(core): patch `/users/:userId/password`
3-before-update-pwd
3-update-pwd
3-after-update-pwd

feat(core): patch `/users/:userId/username`
4-username-before-udpate
4-username-invalid
4-username-update

feat(core): patch `/users/:userId/primary-email`
5-primary-email-before
5-primary-email-exist
5-primary-emial-update

feat(core): patch `/users/:userId/primary-phone`
6-get-phone-before
6-phone-updated

### CODE REVIEW

File: packages/core/src/routes/admin-user.ts [hunk 1]
```diff
@@ -122,10 +125,146 @@ export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
     }
   );
 
+  router.patch(
+    '/users/:userId/password',
+    koaGuard({
+      params: object({ userId: string() }),
+      body: object({ password: string().min(6) }),
```
Comment (darcyYe) [inline packages/core/src/routes/admin-user.ts:132]: maybe add comments to demonstrate the rules of setting a proper/valid password?
the rule can be the length of the password is greater than or equal to six for now, but can be customized in the future or per users' own preferences.

File: packages/core/src/routes/admin-user.ts [hunk 2]
```diff
@@ -122,10 +125,146 @@ export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
...
+    }),
+    async (ctx, next) => {
+      const {
+        params: { userId },
+        body: { password },
+      } = ctx.guard;
+
+      await findUserById(userId);
+
+      const { passwordEncrypted, passwordEncryptionSalt } = encryptUserPassword(userId, password);
```
Comment (darcyYe) [inline packages/core/src/routes/admin-user.ts:142]: should we also expose 'passwordEncryptionMethod'?

File: packages/core/src/routes/admin-user.ts [hunk 3]
```diff
@@ -122,10 +125,146 @@ export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
...
+        passwordEncrypted,
+      });
+      ctx.body = pick(user, ...userInfoSelectFields);
+
+      return next();
+    }
+  );
+
+  router.patch(
+    '/users/:userId/username',
```
Comment (darcyYe) [inline packages/core/src/routes/admin-user.ts:155]: is this use case necessary? I can hardly find a use case like this.

File: packages/core/src/routes/admin-user.ts [hunk 4]
```diff
@@ -122,10 +125,146 @@ export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
...
+
+      return next();
+    }
+  );
+
+  router.patch(
+    '/users/:userId/primary-email',
+    koaGuard({
+      params: object({ userId: string() }),
+      body: object({ primaryEmail: string().email() }),
```
Comment (darcyYe) [inline packages/core/src/routes/admin-user.ts:188]: i remember that 'string().email()' still leads to 'string()' type check. Maybe use emailRegEx instead (in this case, need to refactor the check logic a little)?
image

File: packages/core/src/routes/admin-user.ts [hunk 5]
```diff
@@ -122,10 +125,146 @@ export default function adminUserRoutes<T extends AuthedRouter>(router: T) {
...
+
+      return next();
+    }
+  );
+
+  router.patch(
+    '/users/:userId/primary-phone',
+    koaGuard({
+      params: object({ userId: string() }),
+      body: object({ primaryPhone: string() }),
```
Comment (darcyYe) [inline packages/core/src/routes/admin-user.ts:218]: same as the previous comment.

**Review by darcyYe — CHANGES_REQUESTED** [review]

File: packages/core/src/routes/admin-user.ts (hunk 2 above)
Comment (xiaoyijun) [reply packages/core/src/routes/admin-user.ts]: updated

File: packages/core/src/routes/admin-user.ts (hunk 3 above)
Comment (xiaoyijun) [reply packages/core/src/routes/admin-user.ts]: updated

File: packages/core/src/routes/admin-user.ts (hunk 1 above)
Comment (darcyYe) [reply packages/core/src/routes/admin-user.ts:132]: good

File: packages/core/src/routes/admin-user.ts (hunk 4 above)
Comment (xiaoyijun) [reply packages/core/src/routes/admin-user.ts]: good call. updated

File: packages/core/src/routes/admin-user.ts (hunk 5 above)
Comment (xiaoyijun) [reply packages/core/src/routes/admin-user.ts:194]: updated

**Pull request merged**